design-proposals/2022-10-27-rendering-modularization.md

```diff
@@ -0,0 +1,508 @@
+- Set a uniform of a given type and handle.
+    Perhaps just for OpenGL.
+- Add a new uniform struct of a given handle.
+    Metal doesn’t really do the single named uniform thing like OpenGL does.  It’s better and more efficient to hand over a full struct.
+- Get the list of Drawables that represent everything we’ve built up to this point.
+- Set the emitted drawables up for toolkit specific requests like transitions.
+
+Each SDK will need its own subclass of Builder.
+
+The shared Drawable super-class would contain core functionality across all platforms.  Some of this will be related to things like transitions and tracking tile logic.  Some may be related to data controlled rendering, but more on that later.
+
+In any case the Drawable acts as a basic container that is handed around between the parts of the system that build things and the part that renders things.
+
+Each SDK specific Drawable subclass will do things like:
+- Upload/Bind their data to the SDK.  <br>Ideally this can happen on non-rendering threads, but you never know.
+    This includes figuring out what the shader is asking for and providing it, or convincing defaults.  The way MapLibre Native does this now is a bit static, with templates.  Clever, but perhaps too clever.  It’s okay to wire things up dynamically.
+- Draw directly.   <br>OpenGL is big on this.  You have to set things up, draw, then tear them down.
+    It has the virtue of being simple.  If you’re making changes, you just do them directly.
+- Draw indirectly.   <br>Metal does this.  You add your data to a command buffer where it may be drawn for many frames.
+    This is the fastest and best way to do things, but there are updates you must make between frames.  Obviously, the overall map state needs to be updated, otherwise you’re just redrawing in the same way each time.  Sometimes Drawable specific state changes too and data driven rendering will come into play here.
+- Update for frame.   <br>If you’re drawing indirectly, this is all the state that must be changed for a given frame.
+    Some of this is shared, like the map state (e.g. the matrix controlling positioning, lighting and so forth).
+    Some can be specific to a particular Drawable.  Data driven visuals may fall into this category depending on what the data is.<br>
+    Just to make this even more fun, modern renderers are going to have multiple frames in flight at any given time.  So you can’t just keep one set of values and update them periodically.  My preferred approach is to put memory copy commands into a command buffer with guard logic between them.  In Metal, anyway.  Vulkan will have its own way to do that and OpenGL just doesn't.  Well, there's probably an extension somewhere that does, but it's very hard to use and sparsely supported.
+- Tear down their data.   <br>Pretty simple for OpenGL, but with Metal when you’re using heaps (and you should) you actually want another thread to do this.  Thus it’s SDK specific.
+
```

Comment:
Like a lot of extentions it points toward where GPUs went, but is difficult to use.

What you're writing here is a proposal for an OpenGL optimization PR.  I love that idea!  You should do it!

Reply:
> It's not just OpenGL; these features exist in most APIs (and - pipeline management aside - closely resemble GPU hardware on many platforms), so it will make sense to respect these features in the renderer modularization in my opinion.

That's a completely different approach from what we're proposing here.  Let's regroup down below in the comments to discuss this.